tools/QuietBackgroundProcesses/DevHome.QuietBackgroundProcesses.ElevatedServer/main.cpp

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) Microsoft Corporation and Contributors
+// Licensed under the MIT license.
+
+#include <pch.h>
+
+#include <functional>
+#include <memory>
+#include <mutex>
+
+#include <wrl/client.h>
+#include <wrl/wrappers/corewrappers.h>
+#include <wrl/implements.h>
+#include <wrl/module.h>
+#include <wil/result_macros.h>
+#include <wil/token_helpers.h>
+#include <wil/win32_helpers.h>
+
+#include <objbase.h>
+#include <roregistrationapi.h>
+
+#include <DevHome.QuietBackgroundProcesses.QuietBackgroundProcessesManager.h>
+#include "QuietState.h"
+
+using namespace Microsoft::WRL;
+using namespace Microsoft::WRL::Wrappers;
+
+using unique_hstring_array_ptr = wil::unique_any_array_ptr<HSTRING, wil::cotaskmem_deleter, wil::function_deleter<decltype(::WindowsDeleteString), ::WindowsDeleteString>>;
+
+std::mutex g_finishMutex;
+std::condition_variable g_finishCondition;
+bool g_lastInstanceOfTheModuleObjectIsReleased;
+
+bool IsTokenElevated(HANDLE token)
+{
+    auto mandatoryLabel = wil::get_token_information<TOKEN_MANDATORY_LABEL>(token);
+    LONG levelRid = static_cast<SID*>(mandatoryLabel->Label.Sid)->SubAuthority[0];
+    return levelRid == SECURITY_MANDATORY_HIGH_RID;
+}
+
+wil::unique_ro_registration_cookie RegisterWinrtClasses(_In_ PCWSTR serverName, std::function<void()> objectsReleasedCallback)
+{
+    Module<OutOfProc>::Create(objectsReleasedCallback);
+
+    // Get module classes
+    unique_hstring_array_ptr classes;
+    THROW_IF_FAILED(RoGetServerActivatableClasses(HStringReference(serverName).Get(), &classes, reinterpret_cast<DWORD*>(classes.size_address())));
+
+    // Creation callback
+    PFNGETACTIVATIONFACTORY callback = [](HSTRING name, IActivationFactory** factory) -> HRESULT {
+        RETURN_HR_IF(E_UNEXPECTED, wil::compare_string_ordinal(WindowsGetStringRawBuffer(name, nullptr), L"DevHome.QuietBackgroundProcesses.QuietBackgroundProcessesManager", true) != 0);
+
+        auto manager = winrt::make<winrt::DevHome::QuietBackgroundProcesses::factory_implementation::QuietBackgroundProcessesManager>();
+        manager.as<winrt::Windows::Foundation::IActivationFactory>();
+        *factory = static_cast<IActivationFactory*>(winrt::detach_abi(manager));
+        return S_OK;
+    };
+
+    // Register
+    wil::unique_ro_registration_cookie registrationCookie;
+    PFNGETACTIVATIONFACTORY callbacks[1] = { callback };
+    THROW_IF_FAILED(RoRegisterActivationFactories(classes.get(), callbacks, static_cast<UINT32>(classes.size()), &registrationCookie));
+    return registrationCookie;
+}
+
+//int _cdecl wmain(int argc, __in_ecount(argc) PWSTR wargv[])
+int CALLBACK wWinMain(_In_ HINSTANCE, _In_ HINSTANCE, _In_ LPWSTR wargv, _In_ int)

```

Comment:
Just another question, is the C++ used in this PR needed? Could this have been done via C# instead?